### DISCUSSION TITLE
Web Components/Custom Elements

### DISCUSSION THREAD

**Discussion by seanaguinaga** [remix-run/remix#1879] [Q&A] [answered]
Hello

I get this weird error when I use custom elements with Remix:
<img width="1191" alt="Screen Shot 2022-02-10 at 14 22 07" src="https://user-images.githubusercontent.com/15254340/153506753-151001e0-7abd-42d9-9e93-76645c45a52c.png">

I define them in Next.js with this:

<img width="450" alt="Screen Shot 2022-02-10 at 14 24 09" src="https://user-images.githubusercontent.com/15254340/153506834-0389b2df-f0e6-4589-8c44-b7447022f2ce.png">

And can use them anywhere in Next.js just fine

**Where should I define the elements?**

In entry.client?

<img width="477" alt="Screen Shot 2022-02-10 at 14 24 48" src="https://user-images.githubusercontent.com/15254340/153506953-ba8c2fef-555f-4835-bae6-7d248d9a954e.png">

And second

**What does the error message mean and how can I fix it?**

**Comment by sergiodxa** [comment]
That error looks related to what that custom element is doing, you passed a URL of remix.run website so it's probably trying to do a request?

For the question on where to call defineCustomElement, you don't need to create App and call it inside an effect, entry.client will only run client-side so you can call that function directly in the body before calling hydrate.

```ts
defineCustomElements(window);
hydrate(<RemixBrowser />, document);
```

**Comment by rossipedia** [comment, marked as answer]
```js
useEffect(() => {
  ionDefineCustomElements(window);
});
```

This feels wrong... this will call `ionDefineCustomElements` every time the app renders. Also, this will define those elements _after_ the entire DOM for the app has been built, which presumably includes some of the custom elements being built.

Unless I'm mistaken (which is entirely possible), you want them defined before bootstrapping the app:

```diff
let App = () => {
-   useEffect(() => {
-     defineCustomElements(window);
-   });

  return <RemixBrowser />;
};

+ defineCustomElements(window);
hydrate(<App />, document);
```

**Comment by seanaguinaga** [comment]
Got it on where to put the defineCustomElements!

Now for the error:


![](https://user-images.githubusercontent.com/15254340/153506753-151001e0-7abd-42d9-9e93-76645c45a52c.png)


Look's like remix sees the custom element tag as something to import? Not just as markup to render?

**Reply by ryanflorence** [reply]
Remix is just React here, we don't do anything special. Your errors says a resource was not found. Figure out which resource that is and why it's not found I think.

**Reply by zelia-6385** [reply]
Have the same problem remix + ionic/core. If put defineCustomElements(window) in useEffect (in root or entry.client.jsx), I get the same error. If try to put defineCustomElements(window) in entry.client.jsx without useEffect (as described in this article), I get the error of hydration. Was problem resolved for you? Or nothing helped?
![image](https://user-images.githubusercontent.com/43894060/188110699-cd40b12b-23f5-4b4c-9448-edfa7f55bc64.png)
![image](https://user-images.githubusercontent.com/43894060/188110764-b9586d1a-7cc5-4a5e-abef-103cab43f9bc.png)

**Reply by zelia-6385** [reply]
> Remix is just React here, we don't do anything special. Your errors says a resource was not found. Figure out which resource that is and why it's not found I think.

I've dowloaded project with NextJS and started it. Everything Ok with defineCustomElements. Thete are few screens... Are you shure, that it is a stencil problem described above? I added to package.json of Remix project the same versions of react, react-dom and @ionic/core and got the same error - nothing changes finally. As I understand NextJs works, Remix doesn't work
![image](https://user-images.githubusercontent.com/43894060/188318856-d05de3e2-7ae6-4a77-b585-85271f1f87ef.png)
![image](https://user-images.githubusercontent.com/43894060/188318869-578600d7-18b1-4dfa-9eaa-224ab7d7f157.png)
![image](https://user-images.githubusercontent.com/43894060/188318892-68893cc0-55d1-48e5-8db9-16fd4b59ca1b.png)

**Reply by zelia-6385** [reply]
> Remix is just React here, we don't do anything special. Your errors says a resource was not found. Figure out which resource that is and why it's not found I think.

https://github.com/vercel/next.js/tree/canary/examples/with-ionic-typescript - link to NextJs project
